File: 01-Programming-Languages/PHP/Algorithms/DynamicProgramming/fibonacci.php
<?php

/*
Problem: Fibonacci Number
Description: Compute the nth Fibonacci number using dynamic programming (bottom-up).

Approach:
- Use iterative DP with constant space
- Base cases: fib(0) = 0, fib(1) = 1
- Build up to n by tracking only the last two values

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input:  n = 10
Output: 55
*/

function fib(int $n): int {
    if ($n <= 1) return $n;
    $prev2 = 0;
    $prev1 = 1;
    for ($i = 2; $i <= $n; $i++) {
        $curr = $prev1 + $prev2;
        $prev2 = $prev1;
        $prev1 = $curr;
    }
    return $prev1;
}

for ($i = 0; $i <= 10; $i++) {
    echo "fib($i) = " . fib($i) . "\n";
}

File: 01-Programming-Languages/Scala/Algorithms/DynamicProgramming/Fibonacci.scala
// Problem: Fibonacci Sequence
// Description: Compute the nth Fibonacci number using dynamic programming.
//
// Approach:
// - Use bottom-up DP with two variables
// - Build from base cases F(0)=0, F(1)=1 upward
//
// Time Complexity: O(n)
// Space Complexity: O(1)
//
// Example:
// Input: n = 10
// Output: 55

object Fibonacci {
  def fib(n: Int): Long = {
    if (n <= 1) return n
    var a = 0L; var b = 1L
    for (_ <- 2 to n) {
      val temp = a + b
      a = b
      b = temp
    }
    b
  }

  def main(args: Array[String]): Unit = {
    val n = 10
    println(s"Fibonacci($n) = ${fib(n)}")
    for (i <- 0 to 10) {
      println(s"F($i) = ${fib(i)}")
    }
  }
}
